src/main/java/net/neoforged/neoforge/registries/RegistryObject.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.registries;
+
+import java.util.Optional;
+import java.util.function.Supplier;
+import net.minecraft.core.Holder;
+import net.minecraft.core.Registry;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+
+/**
+ * Superclass of {@link DeferredHolder} which provides access to the underlying object, its bound status, and its key
+ * without providing the functionality of {@link Holder}.
+ * <p>
+ * Users may type-erase their {@link DeferredHolder} fields to this class if they do not need the functionality of {@link Holder}.
+ * 
+ * @param <T> The type of the target object
+ * 
+ * @see {@link DeferredHolder} for the implementation of this class
+ */
+public interface RegistryObject<T> extends Supplier<T> {
+    /**
+     * {@return true if the underlying object is available}
+     */
+    boolean isBound();
+
+    /**
+     * Gets the target object if this registry object {@linkplain #isBound() is bound}.
+     *
+     * @throws IllegalStateException If the backing registry is unavailable.
+     * @throws NullPointerException  If the underlying Holder has not been populated (the target object is not registered).
+     */
+    @Override
+    T get();
+
+    /**
+     * If this registry object {@linkplain #isBound() is bound}, returns an optional containing the target, or an empty optional if unbound.
+     */
+    Optional<T> asOptional();
+
+    /**
+     * {@return the resource location of the target object}
+     */
+    ResourceLocation getId();
+
+    /**
+     * {@return the resource key of the target object}
+     */
+    ResourceKey<? super T> getKey();
```

Comment:
I don't like trying to turn the `T` back into `R`.

Reply:
This form provides the same signature with the existing `ResourceKey<R> getKey` on DH